File: packages/vscode-ext/src/engine-client.ts
import * as http from 'http';

export interface EngineHealth { ok: boolean; version: string }
export interface EngineJobResponse { jobId: string; status: string }

function request<T>(port: number, method: string, path: string, body?: unknown): Promise<T> {
  return new Promise((resolve, reject) => {
    const payload = body ? JSON.stringify(body) : undefined;
    const options: http.RequestOptions = {
      hostname: '127.0.0.1',
      port,
      path,
      method,
      headers: {
        'Content-Type': 'application/json',
        ...(payload ? { 'Content-Length': String(Buffer.byteLength(payload)) } : {}),
      },
    };

    const req = http.request(options, (res) => {
      let data = '';
      res.on('data', (chunk: Buffer) => { data += chunk.toString(); });
      res.on('end', () => {
        try {
          const parsed = JSON.parse(data) as T;
          resolve(parsed);
        } catch {
          reject(new Error(`Invalid JSON response from engine: ${data.slice(0, 200)}`));
        }
      });
    });

    req.on('error', reject);
    req.setTimeout(30_000, () => { req.destroy(new Error('Engine request timed out')); });

    if (payload) req.write(payload);
    req.end();
  });
}

export class EngineClient {
  constructor(private readonly port: number) {}

  health(): Promise<EngineHealth> {
    return request<EngineHealth>(this.port, 'GET', '/health');
  }

  getConfig(): Promise<unknown> {
    return request<unknown>(this.port, 'GET', '/config');
  }

  testConnection(): Promise<{ healthy: boolean; latencyMs: number; error?: string }> {
    return request(this.port, 'POST', '/config/test');
  }

  listCrystals(): Promise<unknown[]> {
    return request<unknown[]>(this.port, 'GET', '/crystals');
  }

  syncCrystals(): Promise<{ success: boolean; message: string }> {
    return request(this.port, 'POST', '/crystals/sync');
  }

  startAnalysis(repoPath: string, outputDir?: string): Promise<EngineJobResponse> {
    return request<EngineJobResponse>(this.port, 'POST', '/analyze', { repoPath, outputDir });
  }

  getJob(jobId: string): Promise<unknown> {
    return request<unknown>(this.port, 'GET', `/jobs/${jobId}`);
  }

  pollJob(
    jobId: string,
    onProgress: (job: unknown) => void,
    intervalMs = 800,
  ): Promise<unknown> {
    return new Promise((resolve, reject) => {
      const poll = () => {
        this.getJob(jobId).then(job => {
          const j = job as { progress?: { status?: string }; error?: string };
          onProgress(job);
          const status = j.progress?.status ?? '';
          if (status === 'complete') { resolve(job); return; }
          if (status === 'failed') { reject(new Error(j.error ?? 'Analysis failed')); return; }
          setTimeout(poll, intervalMs);
        }).catch(reject);
      };
      poll();
    });
  }
}

File: packages/jetbrains/src/main/kotlin/dev/geodesic/plugin/engine/EngineClient.kt
package dev.geodesic.plugin.engine

import com.google.gson.Gson
import com.google.gson.JsonObject
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

data class HealthResult(val ok: Boolean, val version: String)
data class ConnectionResult(val healthy: Boolean, val latencyMs: Long, val error: String?)
data class CrystalEntry(val id: String, val name: String, val fitness: Double)
data class SyncResult(val success: Boolean, val message: String)
data class StartJobResult(val jobId: String, val status: String)

data class JobProgress(
    val status: String,
    val stage: String,
    val phiZoneCount: Int,
    val crystalMatch: String?,
    val crystalMatchScore: Double?
)

data class JobResult(
    val jobId: String,
    val progress: JobProgress,
    val result: AnalysisResult?,
    val error: String?
)

data class AnalysisResult(
    val synthesis: SynthesisData,
    val artifactPaths: ArtifactPaths,
    val fingerprint: String
)

data class SynthesisData(
    val gapReport: GapReport,
    val architectureMapMarkdown: String,
    val skillFile: SkillFileMeta
)

data class SkillFileMeta(val meta: RepoMeta, val phiZones: List<PhiZone>)
data class RepoMeta(val repo: String, val analyzedAt: String, val analysisDurationMs: Long)
data class PhiZone(
    val file: String,
    val lineStart: Int,
    val lineEnd: Int,
    val phiFieldCount: Int,
    val protectionMissing: List<String>
)

data class GapReport(
    val repoName: String,
    val overallScore: Int,
    val overallGrade: String,
    val dimensions: List<Dimension>,
    val uncertainDetections: List<UncertainDetection>
)

data class Dimension(
    val dimension: String,
    val score: Int,
    val grade: String,
    val active: Boolean,
    val findings: List<Finding>
)

data class Finding(
    val severity: String,
    val description: String,
    val file: String,
    val lineStart: Int,
    val lineEnd: Int,
    val detail: String,
    val fix: String,
    val deduction: Int
)

data class UncertainDetection(
    val file: String,
    val lineStart: Int,
    val lineEnd: Int,
    val trigger: String,
    val confidencePct: Int,
    val recommendedAction: String
)

data class ArtifactPaths(
    val architectureMap: String,
    val skillFileJson: String,
    val skillFileMd: String,
    val gapReport: String
)

class EngineClient(private val port: Int) {
    private val gson = Gson()
    private val http = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(10))
        .build()

    fun health(): HealthResult {
        val body = get("/health")
        val obj = gson.fromJson(body, JsonObject::class.java)
        return HealthResult(
            ok = obj.get("ok")?.asBoolean ?: false,
            version = obj.get("version")?.asString ?: ""
        )
    }

    fun testConnection(): ConnectionResult {
        val body = post("/config/test", "{}")
        val obj = gson.fromJson(body, JsonObject::class.java)
        return ConnectionResult(
            healthy = obj.get("healthy")?.asBoolean ?: false,
            latencyMs = obj.get("latencyMs")?.asLong ?: 0L,
            error = obj.get("error")?.asString
        )
    }

    fun listCrystals(): List<CrystalEntry> {
        val body = get("/crystals")
        val arr = gson.fromJson(body, Array<JsonObject>::class.java)
        return arr.map { o ->
            CrystalEntry(
                id = o.get("crystalId")?.asString ?: "",
                name = o.get("name")?.asString ?: "",
                fitness = o.get("fitness")?.asDouble ?: 0.0
            )
        }
    }

    fun syncCrystals(): SyncResult {
        val body = post("/crystals/sync", "{}")
        val obj = gson.fromJson(body, JsonObject::class.java)
        return SyncResult(
            success = obj.get("success")?.asBoolean ?: false,
            message = obj.get("message")?.asString ?: ""
        )
    }

    fun startAnalysis(repoPath: String): StartJobResult {
        val payload = """{"repoPath":${gson.toJson(repoPath)}}"""
        val body = post("/analyze", payload)
        val obj = gson.fromJson(body, JsonObject::class.java)
        return StartJobResult(
            jobId = obj.get("jobId")?.asString ?: "",
            status = obj.get("status")?.asString ?: ""
        )
    }

    fun getJob(jobId: String): JobResult {
        val body = get("/jobs/$jobId")
        return gson.fromJson(body, JobResult::class.java)
    }

    fun pollJob(
        jobId: String,
        onProgress: (JobProgress) -> Unit,
        intervalMs: Long = 1500L,
        timeoutMs: Long = 600_000L
    ): JobResult {
        val deadline = System.currentTimeMillis() + timeoutMs
        while (System.currentTimeMillis() < deadline) {
            val job = getJob(jobId)
            onProgress(job.progress)
            val status = job.progress.status
            if (status == "complete") return job
            if (status == "failed") throw RuntimeException("Analysis failed: ${job.error ?: "unknown error"}")
            Thread.sleep(intervalMs)
        }
        throw RuntimeException("Analysis timed out after ${timeoutMs / 1000}s")
    }

    private fun get(path: String): String {
        val req = HttpRequest.newBuilder()
            .uri(URI.create("http://localhost:$port$path"))
            .timeout(Duration.ofSeconds(30))
            .GET()
            .build()
        val resp = http.send(req, HttpResponse.BodyHandlers.ofString())
        if (resp.statusCode() !in 200..299) {
            throw RuntimeException("Engine returned ${resp.statusCode()} for GET $path: ${resp.body()}")
        }
        return resp.body()
    }

    private fun post(path: String, jsonBody: String): String {
        val req = HttpRequest.newBuilder()
            .uri(URI.create("http://localhost:$port$path"))
            .timeout(Duration.ofSeconds(30))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
            .build()
        val resp = http.send(req, HttpResponse.BodyHandlers.ofString())
        if (resp.statusCode() !in 200..299) {
            throw RuntimeException("Engine returned ${resp.statusCode()} for POST $path: ${resp.body()}")
        }
        return resp.body()
    }
}
